src/util.js

```diff
@@ -152,39 +152,49 @@ export function createRegExpRestore () {
     return function() {
         // Now we've snapshotted some properties, escape the lastMatch string
         let esc = /[.?*+^$[\]\\(){}|-]/g,
-            lm = regExpCache.lastMatch.replace(esc, '\\$&'),
-            reg = new List();
+            lastMatch = regExpCache.lastMatch.replace(esc, '\\$&'),
+            exprStr = '';
 
         // If any of the captured strings were non-empty, iterate over them all
         if (has) {
             for (let i = 1; i <= 9; i++) {
                 let m = regExpCache['$'+i];
 
                 // If it's empty, add an empty capturing group
-                if (!m)
-                    lm = '()' + lm;
-
+                if (!m) {
+                  exprStr += '(';
+                  lastMatch = ')' + lastMatch;
+                }
                 // Else find the string in lm and escape & wrap it to capture it
                 else {
                     m = m.replace(esc, '\\$&');
-                    lm = lm.replace(m, '(' + m + ')');
+                    exprStr += lastMatch.substring(0, lastMatch.indexOf(m)) + '(';
+                    lastMatch = m + ')' + lastMatch.substring(lastMatch.indexOf(m) + m.length);
                 }
-
-                // Push it to the reg and chop lm to make sure further groups come after
-                arrPush.call(reg, lm.slice(0, lm.indexOf('(') + 1));
-                lm = lm.slice(lm.indexOf('(') + 1);
             }
         }
 
-        let exprStr = arrJoin.call(reg, '') + lm;
+        exprStr += lastMatch;
 
         // Shorten the regex by replacing each part of the expression with a match
         // for a string of that exact length.  This is safe for the type of
         // expressions generated above, because the expression matches the whole
         // match string, so we know each group and each segment between capturing
         // groups can be matched by its length alone.
-        exprStr = exprStr.replace(/(\\\(|\\\)|[^()])+/g, (match) => {
```

Comment:
@caridy: Just had a play around with implementing this bit of functionality via iteration rather than the existing regex-based approach.  You can see the changes [here](https://github.com/mtlewis/Intl.js/compare/bug/231/fix-regexp-restore-paren-handling...mtlewis:bug/231/iteration-based-regex-shortening?expand=1).  The iteration-based approach didn't turn out to be as clean as I was hoping it would be, so I'd propose we leave the regex-based approach in place.